### PR TITLE
Model comparison v2

### PR DESCRIPTION
**Pull request by tolga-b**
Second version of model comparison for what-if-tool plugin.

**Motivation for features / changes**
PR #1589 added the ability to load two models and compare their inference scores in Facets Dive panel for binary classification and Datapoint Editor tab for classification.
This PR extends the model comparison mode to Global/Partial PDP, Fairness and Performance panels. It also adds comparing regression and multi-class models in all supported panels.
All model prediction related features in facets dive (Inference Value, Inference Error etc.) now also have multiple model support.

The only mode that does not support multiple models is closest counterfactual button which still computes its results on model 1.

**Technical description of changes**
Some variables that used to handle single models became arrays indexed by model number.
Some string keys for facets features now have model number appended to them.

Add the ability to compare two models them for Global PDP, PDP plots:
Now the Global PDP and PDP plots show one line per model per label in the same chart.
(Previously they only showed one line per label.)

Performance and Fairness panel:
Show one ROC curve per model within the same chart. (Previously showed ROC curve for single model.)
Show one threshold and confusion matrix per model. (Previously showed threshold and confusion matrix for single model.)

Add the ability to load two models for regression:
Facets dive features and MSEs in Performance and Fairness panel work for two regression models.

Minor:
- Do not show model number in facets "Inference Score" when only one model is present.
- Categorical PD plots now show the original value as text under the bar (feature name - original value), instead of showing orange bar for original feature and blue for others
- Remove forceUpdateFeatureValueThresholds as it is unnecessary when bound to arrays with "{{}}" in dom-repeat
- Fixes a bug with multi-class button in inference settings panel where clicking on it after the tool is used on a binary model without refreshing the page could throw an exception.

**Screenshots of UI changes**
all features now support two models
all features now support two models

features old
features old

new performance and fairness tab
new performance and fairness tab

old performance and fairness tab
old performance and fairness tab

pdp shows multiple models
pdp shows multiple models

same for multi-class case
same for multi-class case

categorical pdp shows each model as separate bar and labels original value in text
categorical pdp shows each model as separate bar and labels original value in text

all regression features are available in facets
regression features

regression performance shows both models
regression performance

**Detailed steps to verify changes work correctly (as executed by you)**

Tested all offline demos, multi/single model for binary and multi-class models (using the settings panel multi-class option with binary models). 
Tested multiple model regression with age demo by duplicating the first model.
Testing involves visually verifying the functionality of all panels by clicking buttons and verifying the outputs and checking for exceptions in the browser console.

**Alternate designs / implementations considered**

N/A

### CODE REVIEW

File: tensorboard/plugins/interactive_inference/utils/inference_utils.py [hunk 1]
```diff
@@ -509,6 +514,8 @@ def make_json_formatted_for_single_chart(mutant_features,
       was sent for inference. The length of that field should be the same length
       of mutant_features.
     index_to_mutate: The index of the feature being mutated for this chart.
+    model_id: Index of the model corresponding to inference_results_proto to
+      distinguish between multiple models for model comparison
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/utils/inference_utils.py]: this isn't used anywhere in the method?

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 2]
```diff
@@ -1292,14 +1291,16 @@ <h2>Create a distance feature</h2>
                     </button>
                     <iron-collapse class="no-padding-card" id="collapseperformance" opened="{{openedPerformance}}">
                       <template is="dom-if" if="[[isBinaryClassification_(modelType, multiClass)]]">
-                        <template is="dom-repeat" items="[[featureValueThresholds]]">
+                        <template is="dom-repeat" items="[[featureValueThresholds]]" as="featureValueThreshold">
                           <div class="feature-breakdown">
-                            <div class="slider-label">[[getPrintableValue_(item)]]</div>
+                            <div class="slider-label">[[getPrintableValue_(featureValueThreshold)]]</div>
                             <div class="flex">
+                              <template is="dom-repeat" items="{{featureValueThreshold.threshold}}">
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: indent the paper-slider

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 3]
```diff
@@ -1340,9 +1345,12 @@ <h2>Create a distance feature</h2>
                           <div class="feature-breakdown">
                             <div class="slider-label">Threshold for positive classification</div>
                             <div class="flex">
-                              <paper-slider class="slider" editable=true min="0" max="1" step="0.01"
-                                            immediate-value="{{overallThreshold}}" value="[[overallThreshold]]">
-                              </paper-slider>
+                              <template is="dom-repeat" items="{{overallThresholds}}" as="overallThreshold">
+                                <paper-slider class="slider" editable=true min="0" max="1" step="0.01"
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: add label for sliders when more than one model?

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 4]
```diff
@@ -1373,28 +1381,36 @@ <h2>Create a distance feature</h2>
                                     y-axis-formatter="[[getSimpleAxisFormatter(axisPrecision)]]">
                                 </vz-line-chart>
                               </div>
-                              <tf-confusion-matrix counts="[[getConfusionCounts(inferenceStats_, overallThreshold)]]"
-                                                  class="conf-matrix"></tf-confusion-matrix>
-                              </tf-confusion-matrix>
+                              <div>
+                                <template is="dom-repeat" items="[[inferenceStats_]]" index-as="modelInd">
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: label conf matrices when more than one model

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 5]
```diff
@@ -3662,59 +3815,121 @@ <h2>Create a distance feature</h2>
...
         }
 
         // Convert the data into the format expected by vz-bar-chart.
-        const mapped = _.mapValues(data, function(pointsForSeries){
-          return _.map(pointsForSeries, point => convertPoint(point));
+        // Stack models and labels so that each (model,label) is a separate
+        // series
+        self = this;
+        const mapped = {};
+        _.forEach(data, function(modelValues, modelInd){
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: i know the code was like this before, but if you remove function() {} and replace with () => {} here and below, then you don't need the self=this, can just use this.  also if you were to use "self", should be "const self"

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 6]
```diff
@@ -2184,13 +2211,32 @@ <h2>Create a distance feature</h2>
...
+        if (this.getNumberOfModels_() > 1){
+          return inStr + ' ' + (modelNumber + 1);
+        } else {
+          return inStr
+        }
+      },
+
+      getNumberOfModels_: function(){
+        return this.modelName.split(',').length;
+      },
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: its strange that each time we want this we have to do this string split. i think numModels should be a data member that is set/updated at any appropriate time during inference or after model name setting, and then is used without needing recalcuating.

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 7]
```diff
@@ -2281,190 +2333,225 @@ <h2>Create a distance feature</h2>
...
+          for (let modelInd = 0; modelInd < modelThresholds.length;
+              modelInd++) {
+            const inferenceStats = this.inferenceStats_[modelInd];
+            const key = this.createCombinedValueString_(
+              this.featureValueThresholds[i].value,
+              this.featureValueThresholds[i].value2);
+            plotStats.push(inferenceStats.faceted[key]);
+            plotThresholds.push(modelThresholds[modelInd].threshold)
+          }
+        this.plotPr(
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: indenting issue here maybe?

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 8]
```diff
@@ -3012,29 +3152,35 @@ <h2>Create a distance feature</h2>
...
+                    (a, b) => b.score - a.score);
             }
           } else {
             inferenceMap[this.inferences.indices[i]] = [];
-            // TODO(tolgab) Generalize regression to multiple models
-            const result = this.inferences.results[0].regressionResult;
-            inferenceMap[this.inferences.indices[i]][0] =
+            for (let modelNum = 0; modelNum < this.inferences.results.length;
+                 modelNum++){
+            const result = this.inferences.results[modelNum].regressionResult;
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: indenting issue

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 9]
```diff
@@ -3085,19 +3234,26 @@ <h2>Create a distance feature</h2>
                 }
               }
             }
-            item[inferenceLabelStr] = self.labelVocab[item[inferenceValueStr]];
+            for (let modelInd = 0;
+                modelInd < self.getNumberOfModels_();
+                modelInd++) {
+              item[self.strWithModelIndex_(inferenceLabelStr, modelInd)] =
+              self.labelVocab[
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: whitespace

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 10]
```diff
@@ -3109,48 +3265,36 @@ <h2>Create a distance feature</h2>
...
+            item[this.strWithModelIndex_(inferenceValueStr, i)] =
+            this.examplesAndInferences[index].inferences[inferenceIdx][i][
+              0].label;
+          }
           return;
         }
 
         // Binary classifier case.
-        let threshold = this.overallThreshold;
+        let threshold = this.overallThresholds;
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: probably better if this is called "thresholds" as it is a list

File: tensorboard/plugins/interactive_inference/utils/inference_utils.py (hunk 1 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/utils/inference_utils.py]: you are right, removed it.

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 2 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: done

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 5 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: done.

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 6 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: You are right, moved it to variable numModels which gets updated when user clicks Accept in settings pane.

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 7 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: done.

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 8 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: done.

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 9 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: done.

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 10 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: updated.

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 3 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: added.

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 4 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: added.

**Comment by tolga-b** [comment]
Thank you James! Should be all set with the comments.

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 11]
```diff
@@ -2295,194 +2360,231 @@ <h2>Create a distance feature</h2>
         if (!this.inferenceStats_ || !this.visdata ||
             this.selectedLabelFeature == '' ||
             this.examplesAndInferences.length == 0 ||
-            this.examplesAndInferences[0].inferences == null) {
+            this.examplesAndInferences[0].inferences == null ||
+            this.examplesAndInferences[0].inferences[0].length
+              != this.numModels ||
+            (this.inferenceStats_.length != this.numModels &&
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: maybe add comments about this check, as its gotten complex with these new cases

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 12]
```diff
@@ -3025,34 +3187,46 @@ <h2>Create a distance feature</h2>
...
-            this.inferences.indices.length == 0) {
+            this.inferences.indices.length == 0 ||
+            this.inferences.results.length != this.numModels) {
           return;
         }
+        // console.log('nummodels', this.numModels);
+        // console.log('overallthresholds', this.overallThresholds)
+        // console.log('inferencestats', this.inferenceStats_)
+        // console.log('inferences', this.inferences.results)
+        // console.log('examplesandinferences', this.examplesAndInferences)
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: remove commented-out debug  logs

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 13]
```diff
@@ -3680,59 +3863,120 @@ <h2>Create a distance feature</h2>
...
+              title: 'Feature value',
+              evaluate: d => d.value.x,
+          },
+          {
+              title: 'Score',
+              evaluate: d => d.value.y,
+          },
+        ];
+        // If there is more than 1 model, show model number tooltip
+        if (data.length >= 2){
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: nit: space before {

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 14]
```diff
@@ -3680,59 +3863,120 @@ <h2>Create a distance feature</h2>
...
+            keyToChart = labelKey.substring(0, indexOfModelStr);
+            modelToChart =
+              labelKey.substring(indexOfModelStr + 7,
+                labelKey.indexOf(')', indexOfModelStr));
+          }
+          return {'label': keyToChart, 'model': modelToChart};
+      },
+
+      formatChartKey: function(labelKey, modelInd, numberOfModels) {
+          if (numberOfModels == 1){
```
Comment (jameswex) [inline tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: space before {

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 15]
```diff
@@ -467,28 +467,30 @@
...
       }
 
       .roc-y-label {
         position: absolute;
         left: 0;
-        bottom: 80px;
+        bottom: 110px;
         transform: rotate(270deg);
+        font-size: 12px;
+        color: #3c4043;
```
[the remark this answers is not in the document]
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: Mahima added it yesterday.

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 12 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: oops, done!

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 14 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: done.

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 13 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: done.

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html (hunk 11 above)
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: added description.

File: tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html [hunk 16]
```diff
@@ -222,8 +221,9 @@
       }
 
       .pr-line-chart {
-        height: 150px;
-        width: 250px;
+        margin: 6px 6px 6px 6px;
```
[the remark this answers is not in the document]
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/tf_interactive_inference_dashboard/tf-interactive-inference-dashboard.html]: thanks, done.

File: tensorboard/plugins/interactive_inference/utils/inference_utils.py [hunk 17]
```diff
@@ -463,11 +463,14 @@ def chart_for_index(index_to_mutate):
     mutant_features, mutant_examples = make_mutant_tuples(
         example_protos, original_feature, index_to_mutate, viz_params)
 
-    inference_result_proto = platform_utils.call_servo(
-        mutant_examples, serving_bundle)
-    return make_json_formatted_for_single_chart(mutant_features,
+    charts = []
+    for model_id, serving_bundle in enumerate(serving_bundles):
```
[the remark this answers is not in the document]
Comment (tolga-b) [reply tensorboard/plugins/interactive_inference/utils/inference_utils.py]: good catch, removed!

**Comment by tolga-b** [comment]
Did another pass.
Tested all offline demos, multi/single model for binary and multi-class models (using the settings panel multi-class option with binary models). Tested multiple model regression with age demo by duplicating the first model. 

Added binary check for ROC curve plotting in updateInferenceStats. Before, it only checked for the existence of chart, but if someone switches from binary to multi-class using inference settings window without refreshing the page, the chart elements still exist in the dom which causes an error in plotPR (thresholds don't exist for non-binary inferenceStats.faceted[key] is an object not an array so map in plotPR throws an error).

Added "restamp" to binary/multi-class/regression dom-ifs to delete unused chunks when model type changes. There is currently an issue with multi-class selection in inference settings panel where one may trigger a number of thresholds x number of thresholds (100 x 100) confusion matrix when switching from binary to multi-class model without refreshing the page. Without "restamp" this cripples the GUI after that event. With "restamp", the confusion matrix gets cleared after creation so the issue is temporary. Deferring fixing this to a future PR as it is independent from model comparison.

**Pull request merged**